EmotiBitDataParser/src/ofApp.cpp

```diff
@@ -417,230 +434,255 @@ void ofApp::exit() {
+					// ToDo: handle TIMESTAMP_UTC request
+						//|| splitPacket.at(EmotiBitPacket::headerLength).compare(EmotiBitPacket::TypeTag::TIMESTAMP_UTC) == 0)
+					// ToDo: handle multiple request RD messages
+					//|| (splitPacket.size() > EmotiBitPacket::headerLength + 1 &&
+					//(splitPacket.at(EmotiBitPacket::headerLength + 1).compare(EmotiBitPacket::TypeTag::TIMESTAMP_LOCAL) == 0
+					//	|| splitPacket.at(EmotiBitPacket::headerLength + 1).compare(EmotiBitPacket::TypeTag::TIMESTAMP_UTC) == 0)
+					//	)
+				}
 
-		// ToDo: Handle 2^32 timestamp rollover (~49 days)
+				//double now = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+			}
 
-		if (packetHeader.typeTag.compare(EmotiBitPacket::TypeTag::REQUEST_DATA) == 0) {
-			for (uint16_t i = EmotiBitPacket::headerLength; i < splitPacket.size(); i++) {
-				if (splitPacket.at(i).compare(EmotiBitPacket::TypeTag::TIMESTAMP_LOCAL) == 0) {
-					if (allTimestampData.size() > 0 && allTimestampData.back().roundTrip == -1) {
-						// If the previous sync round trip wasn't detected, remove it
-						allTimestampData.pop_back();
+			if (lastRDPacketNumber > -1) { // Only look for timestamps after REQUEST_DATA
+				if (packetHeader.typeTag.compare(EmotiBitPacket::TypeTag::TIMESTAMP_LOCAL) == 0) {
+					// ToDo: Handle TIMESTAMP_UTC
+					// update allTimestampData
+					allTimestampData.back().TS_received = packetHeader.timestamp;
+					if (splitPacket.size() > EmotiBitPacket::headerLength) {
+						allTimestampData.back().TS_sent = splitPacket.at(EmotiBitPacket::headerLength);
 					}
-					allTimestampData.emplace_back();
-					allTimestampData.back().RD = packetHeader.timestamp;
-					//allTimestampData.TS_sent.push_back("");
-					//allTimestampData.TS_received.push_back(0);
-					//allTimestampData.AK.push_back(0);
-					//allTimestampData.roundTrip.push_back(-1);
-					//allTimestampData.c2e.push_back(-1);
-					lastRDPacketNumber = packetHeader.packetNumber;
-
-					auto loggerPtr = loggers.find(timestampFilenameString);
-					if (loggerPtr != loggers.end()) {
-						loggerPtr->second->push("\n" + ofToString(allTimestampData.back().RD) + ",");
+				}
+				if (packetHeader.typeTag.compare(EmotiBitPacket::TypeTag::ACK) == 0) {
+					if (splitPacket.size() > EmotiBitPacket::headerLength) {
+						// verify if AK is for the correct RD
+						if (lastRDPacketNumber == stoi(splitPacket.at(EmotiBitPacket::headerLength))) {
+							allTimestampData.back().AK = packetHeader.timestamp;
+							allTimestampData.back().roundTrip = allTimestampData.back().TS_received - allTimestampData.back().RD;
+							lastRDPacketNumber = -1; // reset the last RD packet number to avoid overwriting from duplicate
+							auto loggerPtr = loggers.find(timestampFilenameString);
+							if (loggerPtr != loggers.end()) {
+								// write TL information into tiemsync file
+								loggerPtr->second->push(ofToString(allTimestampData.back().TS_received) + ",");
```

Comment:
@produceconsumerobot now the timeSyncs file is written only when an `AK` has been received